src/ExponentialPriceDecayNFTAuction.sol

```diff
@@ -0,0 +1,166 @@
+    /// @param currentPrice The current auction price
+    /// @param maxPrice The passed max price the purchaser is willing to pay
+    error MaxPriceTooLow(uint256 currentPrice, uint256 maxPrice);
+
+    /// @notice Creates an auction defined by the passed `auction`
+    /// @dev assumes the nft being sold is already controlled by the auction contract
+    /// @param auction The defintion of the auction
+    /// @return id the id of the auction
+    function startAuction(Auction calldata auction) external returns (uint256 id) {
+        id = auctionID(auction);
+
+        if (auctionStartTime[id] != 0) {
+            revert AuctionExists();
+        }
+
+        auctionStartTime[id] = block.timestamp;
+
+        emit StartAuction(
+            id,
+            auction.auctionAssetID,
+            auction.auctionAssetContract,
+            auction.perPeriodDecayPercentWad,
+            auction.secondsInPeriod,
+            auction.startPrice,
+            auction.paymentAsset
+            );
+    }
+
+    /// @notice purchases the NFT being sold in `auction`, reverts if current auction price exceed maxPrice
+    /// @dev Does not "pull" payment but expects payment to be received after safeTransferFrom call.
+    /// @dev i.e. does not work if msg.sender is EOA.
+    /// @param auction The auction selling the NFT
+    /// @param maxPrice The maximum the caller is willing to pay
+    /// @param data arbitrary data, passed back to caller, along with the amount to pay, in an encoded CallbackInfo
+    function purchaseNFT(Auction calldata auction, uint256 maxPrice, bytes calldata data) external {
+        uint256 id = auctionID(auction);
+        uint256 startTime = auctionStartTime[id];
+
+        if (startTime == 0) {
+            revert InvalidAuction();
+        }
+
+        uint256 beforeBalance = auction.paymentAsset.balanceOf(address(this));
+        uint256 price = _currentPrice(startTime, auction);
+
+        // price changes over time and the price the caller pays
+        // will depend on when their tx is included in a block
+        // passing a max price offers a nice sanity check on
+        // what you pay
+        if (price > maxPrice) {
+            revert MaxPriceTooLow(price, maxPrice);
+        }
+
+        delete auctionStartTime[id];
+
+        // We effectively use this as a callback, via the on receive handler,
+        // allowing the buyer to receive the NFT first and then provide payment,
+        // meaning they could sell the NFT in some arb to provide payment.
+        // TBD if we should have a dedicated callback method that callers should implement.
+        auction.auctionAssetContract.safeTransferFrom(
+            address(this),
+            msg.sender,
+            auction.auctionAssetID,
+            abi.encode(CallbackInfo({price: price, passedData: data}))
+        );
+
+        uint256 received = auction.paymentAsset.balanceOf(address(this)) - beforeBalance;
```

Comment:
what if for whatever reason we received more than `price`? should we refund winner?

Reply:
I do not think we should refund overpayment